chia/data_layer/data_layer_wallet.py

```diff
@@ -736,6 +737,78 @@ async def singleton_removed(self, parent_spend: CoinSpend, height: uint32) -> No
+            unconfirmed_singletons[0].lineage_proof.parent_name
+        )
+        if parent_singleton is None or {parent_singleton.root} != {s.root for s in full_branch if s.confirmed}:
+            root_changed: bool = True
+        else:
+            root_changed = False
+
+        # Regardless of whether the root changed or not, our old state is bad so let's eliminate it
+        # First let's find all of our txs matching our unconfirmed singletons
+        unconfirmed_ids: Set[bytes32] = {s.lineage_proof.parent_name for s in unconfirmed_singletons}
+        relevant_dl_txs: List[TransactionRecord] = []
+        for id in unconfirmed_ids:
+            tx = await self.wallet_state_manager.tx_store.get_transaction_record(id)
+            if tx is not None:
+                relevant_dl_txs.append(tx)
+        # Let's check our standard wallet for fee transactions related to these dl txs
+        all_spends: List[SpendBundle] = [tx.spend_bundle for tx in relevant_dl_txs if tx.spend_bundle is not None]
+        all_removal_ids: Set[bytes32] = {removal.name() for sb in all_spends for removal in sb.removals()}
+        unconfirmed_std_txs: List[
+            TransactionRecord
+        ] = await self.wallet_state_manager.tx_store.get_unconfirmed_for_wallet(self.standard_wallet.id())
+        relevant_std_txs: List[TransactionRecord] = [
+            tx for tx in unconfirmed_std_txs if all_removal_ids & set([c.name() for c in tx.removals])
+        ]
+        # Delete all of the relevant transactions
+        for tx in [*relevant_dl_txs, *relevant_std_txs]:
+            await self.wallet_state_manager.tx_store.delete_transaction_record(tx.name)
+        # Delete all of the unconfirmed singleton records
+        for singleton in unconfirmed_singletons:
+            await self.wallet_state_manager.dl_store.delete_singleton_record(singleton.coin_id)
+
+        if not root_changed:
+            # The root never changed so let's attempt a rebase
+            all_txs: List[TransactionRecord] = []
+            try:
+                for singleton in unconfirmed_singletons:
+                    for tx in relevant_dl_txs:
+                        if {singleton.coin_id} & set([c.name() for c in tx.additions]):
+                            if tx.spend_bundle is not None:
+                                fee = uint64(tx.spend_bundle.fees())
+                            else:
+                                fee = uint64(0)
+
+                            all_txs.extend(await self.create_update_state_spend(launcher_id, singleton.root, fee))
+            except Exception:
+                # Something went wrong so let's delete anything pending that was created
+                for singleton in unconfirmed_singletons:
+                    await self.wallet_state_manager.dl_store.delete_singleton_record(singleton.coin_id)
+            for tx in all_txs:
+                await self.wallet_state_manager.add_pending_transaction(tx)
```

Comment:
Should this really be run when there was an exception in the `try` above?  Or maybe the `except:` block needs a `raise`.